core/micro-services/workflow-core/src/main/scala/edu/uci/ics/amber/core/executor/SourceOperatorExecutor.scala

```diff
@@ -0,0 +1,45 @@
+package edu.uci.ics.amber.core.executor
+
+import edu.uci.ics.amber.core.tuple.{Tuple, TupleLike}
+import edu.uci.ics.amber.storage.model.DatasetFileDocument
+import edu.uci.ics.amber.workflow.PortIdentity
+
+import java.io.{FileInputStream, InputStream}
+
+trait SourceOperatorExecutor extends OperatorExecutor {
+  override def open(): Unit = {}
+
+  override def close(): Unit = {}
+  override def processTupleMultiPort(
+      tuple: Tuple,
+      port: Int
+  ): Iterator[(TupleLike, Option[PortIdentity])] = Iterator()
+
+  override def processTuple(tuple: Tuple, port: Int): Iterator[TupleLike] = Iterator.empty
+
+  def produceTuple(): Iterator[TupleLike]
+
+  override def onFinishMultiPort(port: Int): Iterator[(TupleLike, Option[PortIdentity])] = {
+    // We assume there is only one input port for source operators. The current assumption
+    // makes produceTuple to be invoked on each input port finish.
+    // We should move this to onFinishAllPorts later.
+    produceTuple().map(t => (t, Option.empty))
+  }
+
+  // this function create the input stream accordingly:
+  // - if filePath is set, create the stream from the file
+  // - if fileDesc is set, create the stream via JGit call
+  def createInputStream(filePath: String, datasetFileDocument: DatasetFileDocument): InputStream = {
```

Comment:
Pass an Either. It should not have two parameters.

Reply:
this function is out-of-sync. It has been removed